src/main/java/org/opensearch/knn/index/codec/KNN990Codec/NativeEngineFieldVectorsWriter.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.index.codec.KNN990Codec;
+
+import lombok.Getter;
+import org.apache.lucene.codecs.KnnFieldVectorsWriter;
+import org.apache.lucene.index.DocsWithFieldSet;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.util.InfoStream;
+import org.apache.lucene.util.RamUsageEstimator;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * NativeEngineVectorFieldsWriter is a class that will be used to accumulate all the vectors during ingestion before
+ * lucene does a flush. This class ensures that KNNVectorWriter is free from generics and this class can encapsulate
+ * all the details related to vectors types and docIds.
+ *
+ * @param <T> float[] or byte[]
+ */
+class NativeEngineFieldVectorsWriter<T> extends KnnFieldVectorsWriter<T> {
```

Comment:
I was thinking if it would be better to have KNNVector class and make class look like 
```
class NativeEngineFieldVectorsWriter<KNNVector> extends KnnFieldVectorsWriter<KNNVector> {
```

```
public interface KNNVector{
  // common methods
}
public class KNNFloatVector implement KNNVector{
  private float[];
}
public class KNNByteVector implement KNNVector{
  private byte[];
}
```

Reply:
not able to understand the benefit of this over float[]. Having another object on top of floats means more objects in heap when this class is not providing any other abstraction.

Reply:
1. It can enforce the interface instead of opening up for any object for `NativeEngineFieldVectorsWriter` Currently, we are enforcing it through constructor which might not intuitive.
2. More flexibility. For example, if we need some other method specific to float or byte, current architecture might not support it smoothly but we will have to add if/else inside `NativeEngineFieldVectorsWriter`.
3. For the return value, we don't know what the T will be so we used `?`. The caller have to validate its instance type before handling it.

Reply:
As discussed, having a base interface like KNNVector just over native types creates an overhead. I did the following change and found out that there is ~60byte overhead that we will get per vector if we add any extra abstraction. You can use the `ramBytesUsed` function to get these details.

```
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

package org.opensearch.knn.common;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;
import org.apache.lucene.util.RamUsageEstimator;

@Getter
@EqualsAndHashCode
@ToString
abstract class KNNVector {
    protected int dimension;

    public long ramBytesUsed() {
        return RamUsageEstimator.sizeOf(dimension);
    }
}


/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

package org.opensearch.knn.common;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

@Getter
@EqualsAndHashCode(callSuper = true)
@ToString(callSuper = true)
class KNNFloatVector extends KNNVector {
    private final float[] vector;

    public KNNFloatVector(float[] vector) {
        if (vector == null || vector.length == 0) {
            throw new IllegalStateException("The input vector for KNNFloatVector Object cannot be null");
        }
        this.dimension = vector.length;
        this.vector = vector;
    }
}


/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

package org.opensearch.knn.common;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import org.apache.lucene.util.RamUsageEstimator;

@Getter
@EqualsAndHashCode(callSuper = true)
class KNNByteVector extends KNNVector {
    private static final long SHALLOW_SIZE = RamUsageEstimator.shallowSizeOfInstance(KNNByteVector.class);

    private final byte[] vector;

    public KNNByteVector(byte[] vector) {
        if (vector == null || vector.length == 0) {
            throw new IllegalStateException("The input vector for KNNByteVector Object cannot be null");
        }
        this.dimension = vector.length;
        this.vector = vector;
    }

    @Override
    public long ramBytesUsed() {
        return SHALLOW_SIZE + ramBytesWithoutInstanceSize();
    }

    protected long ramBytesWithoutInstanceSize() {
        return super.ramBytesUsed() + RamUsageEstimator.shallowSizeOf(vector);
    }
}


/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

package org.opensearch.knn.common;

import lombok.EqualsAndHashCode;
import lombok.ToString;
import org.apache.lucene.util.RamUsageEstimator;

@EqualsAndHashCode(callSuper = true)
@ToString(callSuper = true)
class KNNBinaryVector extends KNNByteVector {
    private static final long SHALLOW_SIZE = RamUsageEstimator.shallowSizeOfInstance(KNNBinaryVector.class);

    public KNNBinaryVector(final byte[] vector) {
        super(vector);
        this.dimension = getDimension() * Byte.SIZE;
    }

    @Override
    public long ramBytesUsed() {
        long s = SHALLOW_SIZE;
        return super.ramBytesWithoutInstanceSize() + SHALLOW_SIZE;
    }
}

```

Another thing is NativeEngineFieldWriter generic is limited to codec and not exposed outside. Plus for iterating over vectors I have added concrete classes in this PR: https://github.com/opensearch-project/k-NN/pull/1897/files. This should resolve all the issues.

Based on this I am closing the issue. If we deicide later that we need to add a low level interface will add it.